python/ray/air/tests/test_experiment_restore.py

```diff
@@ -0,0 +1,250 @@
+import time
+import shutil
+import signal
+import subprocess
+import sys
+
+from ray.tune.result_grid import ResultGrid
+from ray.tune.analysis import ExperimentAnalysis
+
+
+_RUN_SCRIPT_FILENAME = "_test_experiment_restore_run.py"
+
+
+def kill_process_if_needed(
+    process: subprocess.Popen, timeout_s: float = 10, poll_interval_s: float = 1.0
+):
+    """Kills a process if it hasn't finished in `timeout_s` seconds.
+    Polls every `poll_interval_s` seconds to check if the process is still running."""
+    kill_timeout = time.monotonic() + timeout_s
+    while process.poll() is None and time.monotonic() < kill_timeout:
+        time.sleep(poll_interval_s)
+    if process.poll() is None:
+        process.terminate()
+
+
+def print_message(message):
+    print("\n")
+    print("=" * 50)
+    print(message)
+    print("=" * 50)
+    print("\n")
+
+
+@pytest.mark.parametrize("runner_type", ["tuner", "trainer"])
+def test_experiment_restore(tmp_path, runner_type):
+    np.random.seed(2023)
+
+    script_path = Path(__file__).parent / _RUN_SCRIPT_FILENAME
+
+    # Args to pass into the script as environment variables
+    exp_name = f"{runner_type}_restore_integration_test"
+    callback_dump_file = tmp_path / f"{runner_type}-callback_dump_file.json"
+    storage_path = tmp_path / "ray_results"
+    if storage_path.exists():
+        shutil.rmtree(storage_path)
+
+    csv_file = str(tmp_path / "dummy_data.csv")
+    dummy_df = pd.DataFrame({"x": np.arange(128), "y": 2 * np.arange(128)})
+    dummy_df.to_csv(csv_file)
+
+    run_started_marker = tmp_path / "run_started_marker"
+
+    time_per_iter_s = 0.5
+    max_concurrent = 2
+
+    if runner_type == "tuner":
+        iters_per_trial = 8
+        num_trials = 8
+    elif runner_type == "trainer":
+        iters_per_trial = 32
+        num_trials = 1
+
+    total_iters = iters_per_trial * num_trials
+
+    env = {
+        "RUNNER_TYPE": runner_type,
+        "STORAGE_PATH": str(storage_path),
+        "EXP_NAME": exp_name,
+        "CALLBACK_DUMP_FILE": str(callback_dump_file),
+        "RUN_STARTED_MARKER": str(run_started_marker),
+        "TIME_PER_ITER_S": str(time_per_iter_s),
+        "ITERATIONS_PER_TRIAL": str(iters_per_trial),
+        "NUM_TRIALS": str(num_trials),
+        "MAX_CONCURRENT_TRIALS": str(max_concurrent),
+        "CSV_DATA_FILE": csv_file,
+    }
+
+    # Pass criteria
+    no_interrupts_runtime = 16.0
+    passing_factor = 1.5
+    passing_runtime = no_interrupts_runtime * passing_factor
+    print_message(
+        "Experiment should finish with a total runtime of\n"
+        f"<= {passing_runtime} seconds."
+    )
+
+    # Variables used in the loop
+    return_code = None
+    total_runtime = 0
+    run_iter = 0
+    progress_history = []
+
+    while total_runtime < passing_runtime:
+        run_started_marker.write_text("", encoding="utf-8")
+
+        run = subprocess.Popen(
+            [sys.executable, script_path], env=env  # , stderr=subprocess.PIPE
```

Comment:
actually why do you want to go the subprocess route?
why not just write those train() and tune() functions as test code here, and simply call the functions?

Reply:
I think this is because we are simulating script interruption by user input here, which is closer to the user behavior if run in a subprocess